core/types/types.go

```diff
@@ -56,57 +50,67 @@ type BlockHeader struct {
 	StateRoot      common.Hash `json:"stateRoot"`
 	ExtrinsicsRoot common.Hash `json:"extrinsicsRoot"`
 	Digest         []byte      `json:"digest"` // any additional block info eg. logs, seal
+	hash           common.Hash
 }
 
-func (bh *BlockHeader) Hash() (common.Hash, error) {
-	enc, err := scale.Encode(bh)
-	if err != nil {
-		return [32]byte{}, err
+// NewBlockHeader creates a new block header and sets its hash field
+func NewBlockHeader(parentHash common.Hash, number *big.Int, stateRoot common.Hash, extrinsicsRoot common.Hash, digest []byte) (*BlockHeader, error) {
+	bh := &BlockHeader{
+		ParentHash:     parentHash,
+		Number:         number,
+		StateRoot:      stateRoot,
+		ExtrinsicsRoot: extrinsicsRoot,
+		Digest:         digest,
 	}
 
-	return common.Blake2bHash(enc)
-}
-
-func (bh *BlockHeader) WithHash() (*BlockHeaderWithHash, error) {
-	enc, err := scale.Encode(bh)
+	hash, err := bh.Hash()
 	if err != nil {
 		return nil, err
 	}
 
-	hash, err := common.Blake2bHash(enc)
-	if err != nil {
-		return nil, err
+	bh.hash = hash
+	return bh, nil
+}
+
+// MustHash returns the hash of the block header
+// If the internal hash field is nil, it hashes the block and sets the hash field.
+// If hashing the header errors, this will panic.
+func (bh *BlockHeader) MustHash() common.Hash {
```

Comment:
Why do we need both `Hash` and `MustHash`?

Reply:
it's so we can use it directly as a hash, without needing a multiple-liner to check err. eg. see where MustHash is used. I could remove it, it's mostly for convenience I suppose.